js/private/js_binary.bzl

```diff
@@ -322,18 +323,24 @@ _NODE_OPTION = """JS_BINARY__NODE_OPTIONS+=(\"{value}\")"""
 def _deprecated_target_tool_path_to_short_path(tool_path):
     return ("../" + tool_path[len("external/"):]) if tool_path.startswith("external/") else tool_path
 
+def _expand_env_if_needed(ctx, value):
+    if ctx.attr.expand_env:
+        return " ".join([expand_variables(ctx, exp, attribute_name = "env") for exp in expand_locations(ctx, value, ctx.attr.data).split(" ")])
+    return value
+
 def _bash_launcher(ctx, nodeinfo, entry_point_path, log_prefix_rule_set, log_prefix_rule, fixed_args, fixed_env, is_windows):
     # Explicitly disable node fs patches on Windows:
     # https://github.com/aspect-build/rules_js/issues/1137
     if is_windows:
         fixed_env = dict(fixed_env, **{"JS_BINARY__PATCH_NODE_FS": "0"})
 
-    envs = []
-    for (key, value) in dicts.add(fixed_env, ctx.attr.env).items():
-        envs.append(_ENV_SET.format(
-            var = key,
-            value = " ".join([expand_variables(ctx, exp, attribute_name = "env") for exp in expand_locations(ctx, value, ctx.attr.data).split(" ")]),
-        ))
+    envs = [
```

Comment:
refactoring this was around 100ms win, but I was already here so might as well

Reply:
Worth it's own PR so we can merge now?

Reply:
It's a bit wordy without the extracted helper. I'd rather just land this together, don't want to create more work for myself rebasing